system/HTTP/IncomingRequest.php

```diff
@@ -395,33 +399,61 @@ public function setPath(string $path, ?App $config = null)
 
         $config ??= $this->config;
 
+        if ($config->baseURL === '') {
+            // @codeCoverageIgnoreStart
+            exit('You have an empty or invalid base URL. The baseURL value must be set in Config\App.php, or through the .env file.');
+            // @codeCoverageIgnoreEnd
+        }
+
         // It's possible the user forgot a trailing slash on their
         // baseURL, so let's help them out.
-        $baseURL = $config->baseURL === '' ? $config->baseURL : rtrim($config->baseURL, '/ ') . '/';
+        $baseURL = rtrim($config->baseURL, '/ ') . '/';
 
-        // Based on our baseURL provided by the developer
-        // set our current domain name, scheme
-        if ($baseURL !== '') {
-            $this->uri->setScheme(parse_url($baseURL, PHP_URL_SCHEME));
-            $this->uri->setHost(parse_url($baseURL, PHP_URL_HOST));
-            $this->uri->setPort(parse_url($baseURL, PHP_URL_PORT));
+        $host = $this->determineHost($config, $baseURL);
 
-            // Ensure we have any query vars
-            $this->uri->setQuery($_SERVER['QUERY_STRING'] ?? '');
+        // Based on $baseURL and $allowedHostnames provided by Config\App,
+        // set our current domain name, scheme.
+        $this->uri->setScheme(parse_url($baseURL, PHP_URL_SCHEME));
+        $this->uri->setHost($host);
+        $this->uri->setPort(parse_url($baseURL, PHP_URL_PORT));
 
-            // Check if the baseURL scheme needs to be coerced into its secure version
-            if ($config->forceGlobalSecureRequests && $this->uri->getScheme() === 'http') {
-                $this->uri->setScheme('https');
-            }
-        } elseif (! is_cli()) {
-            // @codeCoverageIgnoreStart
-            exit('You have an empty or invalid base URL. The baseURL value must be set in Config\App.php, or through the .env file.');
-            // @codeCoverageIgnoreEnd
+        // Update Config\App::$baseURL
+        $uri = new URI($baseURL);
+        $uri->setHost($host);
+        $this->config->baseURL = $uri;
```

Comment:
Without this, all helpers will return "old" URLs, not one of those defined in `allowedHostnames`. That's basically the whole point of all these changes.

I have an app where subdomains and custom domains are used, but I use a more dynamic approach and handle it via Registrars. I would recommend this way, but the only downside is that we can't make it work with one config variable - it's a bit more complicated.

Reply:
We need the current real baseURL with `allowedHostnames` in somewhere.
I think `Config\App::$baseURL` in shared Config object is best, 
because the baseURL is only one in a request and never changes. I don't see any problems.

But if something wrong that I miss, we need to save the current baseURL somewhere else.
Where?

Reply:
I would rather see anything that needs to rely on "URL for the incoming request" to use a component that represents that (IncomingRequest, current_url(), etc). It seems very counterintuitive that our Config values are dependent on infrastructure context.

I get that this is the most straightforward approach to a fix, but I think it pokes a new hole in our architecture that is liable to leak, grow bigger, etc. 

I'm not a fan of the proliferation of global URL helpers we already have but I would prefer something like `current_base_url()` to this.

Reply:
Re: Registrars... I think that is a great solution at the project level, and would encourage it for individuals who need this currently.

Reply:
Indeed, it may seem counter-intuitive that the value of one Config object changes dynamically.
I think it is better that the URI object in IncomingRequest has all the URL related data including the current baseURL.